<!DOCTYPE html>
<html lang='en'>
<head>
<title>0001-xwayland-support-HiDPI-scale.patch - aur.git - AUR Package Repositories</title>
<meta name='generator' content='cgit v1.2.3-56-gcc08'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/static/css/cgit.css'/>
<script type='text/javascript' src='/cgit.js'></script>
<link rel='shortcut icon' href='/static/images/favicon.ico'/>
<link rel='alternate' title='Atom feed' href='https://aur.archlinux.org/cgit/aur.git/atom/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git' type='application/atom+xml'/>
<link rel='vcs-git' href='https://aur.archlinux.org/wlroots-hidpi-xprop-git.git/' title='aur.git Git repository'/>
</head>
<body>
		<div id="archnavbar" class="anb-aur">
			<div id="archnavbarlogo"><h1><a href="/" title="Return to the main page">Arch Linux User Repository</a></h1></div>
			<div id="archnavbarmenu">
				<ul id="archnavbarlist">
					<li id="anb-home"><a href="https://www.archlinux.org/" title="Arch news, packages, projects and more">Home</a></li>
					<li id="anb-packages"><a href="https://www.archlinux.org/packages/" title="Arch Package Database">Packages</a></li>
					<li id="anb-forums"><a href="https://bbs.archlinux.org/" title="Community forums">Forums</a></li>
					<li id="anb-wiki"><a href="https://wiki.archlinux.org/" title="Community documentation">Wiki</a></li>
					<li id="anb-bugs"><a href="https://bugs.archlinux.org/" title="Report and track bugs">Bugs</a></li>
					<li id="anb-security"><a href="https://security.archlinux.org/" title="Arch Linux Security Tracker">Security</a></li>
					<li id="anb-aur"><a href="/" title="Arch Linux User Repository">AUR</a></li>
					<li id="anb-download"><a href="https://www.archlinux.org/download/" title="Get Arch Linux">Download</a></li>
				</ul>
			</div>
		</div><!-- #archnavbar -->
<div id='cgit'><table id='header'>
<tr>
<td class='main'>AUR : wlroots-hidpi-xprop-git.git</td></tr>
<tr><td class='sub'>AUR Package Repositories | click <a href='/pkgbase/wlroots-hidpi-xprop-git/'>here</a> to return to the package base details page</td><td class='sub right'></td></tr></table>
<table class='tabs'><tr><td>
<a href='/cgit/aur.git/?h=wlroots-hidpi-xprop-git'>summary</a><a href='/cgit/aur.git/log/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>log</a><a class='active' href='/cgit/aur.git/tree/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>tree</a><a href='/cgit/aur.git/commit/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>commit</a><a href='/cgit/aur.git/diff/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>diff</a><a href='/cgit/aur.git/stats/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>stats</a></td><td class='form'><form class='right' method='get' action='/cgit/aur.git/log/0001-xwayland-support-HiDPI-scale.patch'>
<input type='hidden' name='h' value='wlroots-hidpi-xprop-git'/><select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='search' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/cgit/aur.git/tree/?h=wlroots-hidpi-xprop-git'>root</a>/<a href='/cgit/aur.git/tree/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>0001-xwayland-support-HiDPI-scale.patch</a></div><div class='content'>blob: 43d76b4137933922b429022bebd952c175b696f3 (<a href='/cgit/aur.git/plain/0001-xwayland-support-HiDPI-scale.patch?h=wlroots-hidpi-xprop-git'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a id='n1' href='#n1'>1</a>
<a id='n2' href='#n2'>2</a>
<a id='n3' href='#n3'>3</a>
<a id='n4' href='#n4'>4</a>
<a id='n5' href='#n5'>5</a>
<a id='n6' href='#n6'>6</a>
<a id='n7' href='#n7'>7</a>
<a id='n8' href='#n8'>8</a>
<a id='n9' href='#n9'>9</a>
<a id='n10' href='#n10'>10</a>
<a id='n11' href='#n11'>11</a>
<a id='n12' href='#n12'>12</a>
<a id='n13' href='#n13'>13</a>
<a id='n14' href='#n14'>14</a>
<a id='n15' href='#n15'>15</a>
<a id='n16' href='#n16'>16</a>
<a id='n17' href='#n17'>17</a>
<a id='n18' href='#n18'>18</a>
<a id='n19' href='#n19'>19</a>
<a id='n20' href='#n20'>20</a>
<a id='n21' href='#n21'>21</a>
<a id='n22' href='#n22'>22</a>
<a id='n23' href='#n23'>23</a>
<a id='n24' href='#n24'>24</a>
<a id='n25' href='#n25'>25</a>
<a id='n26' href='#n26'>26</a>
<a id='n27' href='#n27'>27</a>
<a id='n28' href='#n28'>28</a>
<a id='n29' href='#n29'>29</a>
<a id='n30' href='#n30'>30</a>
<a id='n31' href='#n31'>31</a>
<a id='n32' href='#n32'>32</a>
<a id='n33' href='#n33'>33</a>
<a id='n34' href='#n34'>34</a>
<a id='n35' href='#n35'>35</a>
<a id='n36' href='#n36'>36</a>
<a id='n37' href='#n37'>37</a>
<a id='n38' href='#n38'>38</a>
<a id='n39' href='#n39'>39</a>
<a id='n40' href='#n40'>40</a>
<a id='n41' href='#n41'>41</a>
<a id='n42' href='#n42'>42</a>
<a id='n43' href='#n43'>43</a>
<a id='n44' href='#n44'>44</a>
<a id='n45' href='#n45'>45</a>
<a id='n46' href='#n46'>46</a>
<a id='n47' href='#n47'>47</a>
<a id='n48' href='#n48'>48</a>
<a id='n49' href='#n49'>49</a>
<a id='n50' href='#n50'>50</a>
<a id='n51' href='#n51'>51</a>
<a id='n52' href='#n52'>52</a>
<a id='n53' href='#n53'>53</a>
<a id='n54' href='#n54'>54</a>
<a id='n55' href='#n55'>55</a>
<a id='n56' href='#n56'>56</a>
<a id='n57' href='#n57'>57</a>
<a id='n58' href='#n58'>58</a>
<a id='n59' href='#n59'>59</a>
<a id='n60' href='#n60'>60</a>
<a id='n61' href='#n61'>61</a>
<a id='n62' href='#n62'>62</a>
<a id='n63' href='#n63'>63</a>
<a id='n64' href='#n64'>64</a>
<a id='n65' href='#n65'>65</a>
<a id='n66' href='#n66'>66</a>
<a id='n67' href='#n67'>67</a>
<a id='n68' href='#n68'>68</a>
<a id='n69' href='#n69'>69</a>
<a id='n70' href='#n70'>70</a>
<a id='n71' href='#n71'>71</a>
<a id='n72' href='#n72'>72</a>
<a id='n73' href='#n73'>73</a>
<a id='n74' href='#n74'>74</a>
<a id='n75' href='#n75'>75</a>
<a id='n76' href='#n76'>76</a>
<a id='n77' href='#n77'>77</a>
<a id='n78' href='#n78'>78</a>
<a id='n79' href='#n79'>79</a>
<a id='n80' href='#n80'>80</a>
<a id='n81' href='#n81'>81</a>
<a id='n82' href='#n82'>82</a>
<a id='n83' href='#n83'>83</a>
<a id='n84' href='#n84'>84</a>
<a id='n85' href='#n85'>85</a>
<a id='n86' href='#n86'>86</a>
<a id='n87' href='#n87'>87</a>
<a id='n88' href='#n88'>88</a>
<a id='n89' href='#n89'>89</a>
<a id='n90' href='#n90'>90</a>
<a id='n91' href='#n91'>91</a>
<a id='n92' href='#n92'>92</a>
<a id='n93' href='#n93'>93</a>
<a id='n94' href='#n94'>94</a>
<a id='n95' href='#n95'>95</a>
<a id='n96' href='#n96'>96</a>
<a id='n97' href='#n97'>97</a>
<a id='n98' href='#n98'>98</a>
<a id='n99' href='#n99'>99</a>
<a id='n100' href='#n100'>100</a>
<a id='n101' href='#n101'>101</a>
<a id='n102' href='#n102'>102</a>
<a id='n103' href='#n103'>103</a>
<a id='n104' href='#n104'>104</a>
<a id='n105' href='#n105'>105</a>
<a id='n106' href='#n106'>106</a>
<a id='n107' href='#n107'>107</a>
<a id='n108' href='#n108'>108</a>
<a id='n109' href='#n109'>109</a>
<a id='n110' href='#n110'>110</a>
<a id='n111' href='#n111'>111</a>
<a id='n112' href='#n112'>112</a>
<a id='n113' href='#n113'>113</a>
<a id='n114' href='#n114'>114</a>
<a id='n115' href='#n115'>115</a>
<a id='n116' href='#n116'>116</a>
<a id='n117' href='#n117'>117</a>
<a id='n118' href='#n118'>118</a>
<a id='n119' href='#n119'>119</a>
<a id='n120' href='#n120'>120</a>
<a id='n121' href='#n121'>121</a>
<a id='n122' href='#n122'>122</a>
<a id='n123' href='#n123'>123</a>
<a id='n124' href='#n124'>124</a>
<a id='n125' href='#n125'>125</a>
<a id='n126' href='#n126'>126</a>
<a id='n127' href='#n127'>127</a>
<a id='n128' href='#n128'>128</a>
<a id='n129' href='#n129'>129</a>
<a id='n130' href='#n130'>130</a>
<a id='n131' href='#n131'>131</a>
<a id='n132' href='#n132'>132</a>
<a id='n133' href='#n133'>133</a>
<a id='n134' href='#n134'>134</a>
<a id='n135' href='#n135'>135</a>
<a id='n136' href='#n136'>136</a>
<a id='n137' href='#n137'>137</a>
<a id='n138' href='#n138'>138</a>
<a id='n139' href='#n139'>139</a>
<a id='n140' href='#n140'>140</a>
<a id='n141' href='#n141'>141</a>
<a id='n142' href='#n142'>142</a>
<a id='n143' href='#n143'>143</a>
<a id='n144' href='#n144'>144</a>
<a id='n145' href='#n145'>145</a>
<a id='n146' href='#n146'>146</a>
<a id='n147' href='#n147'>147</a>
<a id='n148' href='#n148'>148</a>
</pre></td>
<td class='lines'><pre><code>From 179258208d98fe3755b414749e9776e96d3b1dad Mon Sep 17 00:00:00 2001
From: lilydjwg &lt;lilydjwg&#64;gmail.com&gt;
Date: Wed, 17 Nov 2021 19:34:58 +0800
Subject: [PATCH 1/3] xwayland: support HiDPI scale

This supports the xorg-xwayland patch at https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/733
<span class="hl kwa">---</span>
 include/xwayland/xwm.h |  2 ++
 xwayland/xwm.c         | 51 +++++++++++++++++++++++++++++++-----------
 2 files changed, 40 insertions(+), 13 deletions(-)

diff --git a/include/xwayland/xwm.h b/include/xwayland/xwm.h
index 3d540522..33ff1478 100644
<span class="hl kwa">--- a/include/xwayland/xwm.h</span>
<span class="hl kwb">+++ b/include/xwayland/xwm.h</span>
&#64;&#64; -88,6 +88,7 &#64;&#64; enum atom_name {
 	DND_ACTION_PRIVATE,
 	NET_CLIENT_LIST,
 	NET_CLIENT_LIST_STACKING,
<span class="hl kwb">+        XWAYLAND_GLOBAL_OUTPUT_SCALE,</span>
 	ATOM_LAST // keep last
 };
 
&#64;&#64; -96,6 +97,7 &#64;&#64; struct wlr_xwm {
 	struct wl_event_source *event_source;
 	struct wlr_seat *seat;
 	uint32_t ping_timeout;
<span class="hl kwb">+        uint32_t scale;</span>
 
 	xcb_atom_t atoms[ATOM_LAST];
 	xcb_connection_t *xcb_conn;
diff --git a/xwayland/xwm.c b/xwayland/xwm.c
index 0c30d8f6..59e9ce06 100644
<span class="hl kwa">--- a/xwayland/xwm.c</span>
<span class="hl kwb">+++ b/xwayland/xwm.c</span>
<span class="hl kwd">&#64;&#64; -19,6 +19,14 &#64;&#64;</span>
 #include &lt;xcb/xfixes.h&gt;
 #include &quot;xwayland/xwm.h&quot;
 
<span class="hl kwb">+static int32_t scale(struct wlr_xwm *xwm, uint32_t val) {</span>
<span class="hl kwb">+	return val * xwm-&gt;scale;</span>
<span class="hl kwb">+}</span>
<span class="hl kwb">+</span>
<span class="hl kwb">+static int32_t unscale(struct wlr_xwm *xwm, uint32_t val) {</span>
<span class="hl kwb">+	return (val + xwm-&gt;scale/2) / xwm-&gt;scale;</span>
<span class="hl kwb">+}</span>
<span class="hl kwb">+</span>
 static const char *const atom_map[ATOM_LAST] = {
 	[WL_SURFACE_ID] = &quot;WL_SURFACE_ID&quot;,
 	[WL_SURFACE_SERIAL] = &quot;WL_SURFACE_SERIAL&quot;,
&#64;&#64; -90,6 +98,7 &#64;&#64; static const char *const atom_map[ATOM_LAST] = {
 	[DND_ACTION_PRIVATE] = &quot;XdndActionPrivate&quot;,
 	[NET_CLIENT_LIST] = &quot;_NET_CLIENT_LIST&quot;,
 	[NET_CLIENT_LIST_STACKING] = &quot;_NET_CLIENT_LIST_STACKING&quot;,
<span class="hl kwb">+	[XWAYLAND_GLOBAL_OUTPUT_SCALE] = &quot;_XWAYLAND_GLOBAL_OUTPUT_SCALE&quot;,</span>
 };
 
 #define STARTUP_INFO_REMOVE_PREFIX &quot;remove: ID=&quot;
&#64;&#64; -962,8 +971,8 &#64;&#64; static void xwm_handle_create_notify(struct wlr_xwm *xwm,
 		return;
 	}
 
<span class="hl kwa">-	xwayland_surface_create(xwm, ev-&gt;window, ev-&gt;x, ev-&gt;y,</span>
<span class="hl kwa">-		ev-&gt;width, ev-&gt;height, ev-&gt;override_redirect);</span>
<span class="hl kwb">+	xwayland_surface_create(xwm, ev-&gt;window, unscale(xwm, ev-&gt;x), unscale(xwm, ev-&gt;y),</span>
<span class="hl kwb">+		unscale(xwm, ev-&gt;width), unscale(xwm, ev-&gt;height), ev-&gt;override_redirect);</span>
 }
 
 static void xwm_handle_destroy_notify(struct wlr_xwm *xwm,
&#64;&#64; -994,10 +1003,10 &#64;&#64; static void xwm_handle_configure_request(struct wlr_xwm *xwm,
 
 	struct wlr_xwayland_surface_configure_event wlr_event = {
 		.surface = surface,
<span class="hl kwa">-		.x = mask &amp; XCB_CONFIG_WINDOW_X ? ev-&gt;x : surface-&gt;x,</span>
<span class="hl kwa">-		.y = mask &amp; XCB_CONFIG_WINDOW_Y ? ev-&gt;y : surface-&gt;y,</span>
<span class="hl kwa">-		.width = mask &amp; XCB_CONFIG_WINDOW_WIDTH ? ev-&gt;width : surface-&gt;width,</span>
<span class="hl kwa">-		.height = mask &amp; XCB_CONFIG_WINDOW_HEIGHT ? ev-&gt;height : surface-&gt;height,</span>
<span class="hl kwb">+		.x = mask &amp; XCB_CONFIG_WINDOW_X ? unscale(xwm, ev-&gt;x) : surface-&gt;x,</span>
<span class="hl kwb">+		.y = mask &amp; XCB_CONFIG_WINDOW_Y ? unscale(xwm, ev-&gt;y) : surface-&gt;y,</span>
<span class="hl kwb">+		.width = mask &amp; XCB_CONFIG_WINDOW_WIDTH ? unscale(xwm, ev-&gt;width) : surface-&gt;width,</span>
<span class="hl kwb">+		.height = mask &amp; XCB_CONFIG_WINDOW_HEIGHT ? unscale(xwm, ev-&gt;height) : surface-&gt;height,</span>
 		.mask = mask,
 	};
 
&#64;&#64; -1012,14 +1021,14 &#64;&#64; static void xwm_handle_configure_notify(struct wlr_xwm *xwm,
 	}
 
 	bool geometry_changed =
<span class="hl kwa">-		(xsurface-&gt;x != ev-&gt;x || xsurface-&gt;y != ev-&gt;y ||</span>
<span class="hl kwa">-		 xsurface-&gt;width != ev-&gt;width || xsurface-&gt;height != ev-&gt;height);</span>
<span class="hl kwb">+		(xsurface-&gt;x != unscale(xwm, ev-&gt;x) || xsurface-&gt;y != unscale(xwm, ev-&gt;y) ||</span>
<span class="hl kwb">+		 xsurface-&gt;width != unscale(xwm, ev-&gt;width) || xsurface-&gt;height != unscale(xwm, ev-&gt;height));</span>
 
 	if (geometry_changed) {
<span class="hl kwa">-		xsurface-&gt;x = ev-&gt;x;</span>
<span class="hl kwa">-		xsurface-&gt;y = ev-&gt;y;</span>
<span class="hl kwa">-		xsurface-&gt;width = ev-&gt;width;</span>
<span class="hl kwa">-		xsurface-&gt;height = ev-&gt;height;</span>
<span class="hl kwb">+		xsurface-&gt;x = unscale(xwm, ev-&gt;x);</span>
<span class="hl kwb">+		xsurface-&gt;y = unscale(xwm, ev-&gt;y);</span>
<span class="hl kwb">+		xsurface-&gt;width = unscale(xwm, ev-&gt;width);</span>
<span class="hl kwb">+		xsurface-&gt;height = unscale(xwm, ev-&gt;height);</span>
 	}
 
 	if (xsurface-&gt;override_redirect != ev-&gt;override_redirect) {
&#64;&#64; -1136,7 +1145,22 &#64;&#64; static void xwm_handle_unmap_notify(struct wlr_xwm *xwm,
 static void xwm_handle_property_notify(struct wlr_xwm *xwm,
 		xcb_property_notify_event_t *ev) {
 	struct wlr_xwayland_surface *xsurface = lookup_surface(xwm, ev-&gt;window);
<span class="hl kwb">+</span>
 	if (xsurface == NULL) {
<span class="hl kwb">+		if (ev-&gt;atom == xwm-&gt;atoms[XWAYLAND_GLOBAL_OUTPUT_SCALE]) {</span>
<span class="hl kwb">+			xcb_get_property_cookie_t cookie = xcb_get_property(xwm-&gt;xcb_conn, 0,</span>
<span class="hl kwb">+					ev-&gt;window, ev-&gt;atom, XCB_ATOM_ANY, 0, 2048);</span>
<span class="hl kwb">+			xcb_get_property_reply_t *reply = xcb_get_property_reply(xwm-&gt;xcb_conn,</span>
<span class="hl kwb">+					cookie, NULL);</span>
<span class="hl kwb">+			if (reply == NULL) {</span>
<span class="hl kwb">+				return;</span>
<span class="hl kwb">+			}</span>
<span class="hl kwb">+			if (reply-&gt;type == XCB_ATOM_CARDINAL) {</span>
<span class="hl kwb">+				xwm-&gt;scale = *(uint32_t*)xcb_get_property_value(reply);</span>
<span class="hl kwb">+			}</span>
<span class="hl kwb">+			free(reply);</span>
<span class="hl kwb">+		}</span>
<span class="hl kwb">+</span>
 		return;
 	}
 
&#64;&#64; -1770,7 +1794,7 &#64;&#64; void wlr_xwayland_surface_configure(struct wlr_xwayland_surface *xsurface,
 	uint32_t mask = XCB_CONFIG_WINDOW_X | XCB_CONFIG_WINDOW_Y |
 		XCB_CONFIG_WINDOW_WIDTH | XCB_CONFIG_WINDOW_HEIGHT |
 		XCB_CONFIG_WINDOW_BORDER_WIDTH;
<span class="hl kwa">-	uint32_t values[] = {x, y, width, height, 0};</span>
<span class="hl kwb">+	uint32_t values[] = {scale(xwm, x), scale(xwm, y), scale(xwm, width), scale(xwm, height), 0};</span>
 	xcb_configure_window(xwm-&gt;xcb_conn, xsurface-&gt;window_id, mask, values);
 
 	// If the window size did not change, then we cannot rely on
&#64;&#64; -2123,6 +2147,7 &#64;&#64; struct wlr_xwm *xwm_create(struct wlr_xwayland *xwayland, int wm_fd) {
 	wl_list_init(&amp;xwm-&gt;pending_startup_ids);
 	xwm-&gt;ping_timeout = 10000;
 
<span class="hl kwb">+	xwm-&gt;scale = 1;</span>
 	xwm-&gt;xcb_conn = xcb_connect_to_fd(wm_fd, NULL);
 
 	int rc = xcb_connection_has_error(xwm-&gt;xcb_conn);
<span class="hl kwa">-- </span>
2.41.0

</code></pre></td></tr></table>
</div> <!-- class=content -->
<div id="footer">
	<p>
		Copyright &copy; 2004-2023 aurweb Development Team &ndash;
		<strong>AUR packages are user produced content. Any use of the provided files is at your own risk.</strong>
	</p>
</div>
</div> <!-- id=cgit -->
</body>
</html>
